src/GitVersionCore/Helpers/LockFile.cs

```diff
@@ -0,0 +1,208 @@
+/*
+MIT License
+
+Copyright 2020 Teroneko
+
+Permission is hereby granted, free of charge, to any person obtaining a copy 
+of this software and associated documentation files (the "Software"), to deal 
+in the Software without restriction, including without limitation the rights to 
+use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies 
+of the Software, and to permit persons to whom the Software is furnished to do 
+so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all 
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER 
+IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+using System;
+using System.IO;
+using System.Threading;
+
+namespace GitVersion.Helpers
+{
+    /// <summary>
+    /// This helper class can lock files.
+    /// </summary>
+    public static class LockFile
+    {
+        public const FileMode DefaultFileMode = FileMode.OpenOrCreate;
+        public const FileAccess DefaultFileAccess = FileAccess.ReadWrite;
+        public const FileShare DefaultFileShare = FileShare.None;
+        public const int DefaultTimeoutInMilliseconds = Timeout.Infinite;
+
+        /// <summary>
+        /// Try to acquire lock on file but only as long the file stream is opened.
+        /// </summary>
+        /// <param name="filePath">The path to file that get locked.</param>
+        /// <param name="fileStream">The locked file as file stream.</param>
+        /// <param name="fileMode">The file mode when opening file.</param>
+        /// <param name="fileAccess">The file access when opening file.</param>
+        /// <param name="fileShare">The file share when opening file</param>
+        /// <returns>If true the lock acquirement was successful.</returns>
+        public static bool TryAcquire(string filePath, out FileStream? fileStream, FileMode fileMode = DefaultFileMode,
+            FileAccess fileAccess = DefaultFileAccess, FileShare fileShare = DefaultFileShare)
+        {
+            filePath = filePath ?? throw new ArgumentNullException(nameof(filePath));
+
+            try
+            {
+                fileStream = File.Open(filePath, fileMode, fileAccess, fileShare);
+                return true;
+            }
+            catch (Exception error) when (error.GetType() == typeof(IOException))
```

Comment:
Can't this simply be written as:

```suggestion
            catch (IOException exception)
```

Reply:
No this is a crucial part of the API. The explanation is, that when you open a FileStream that failes because it was locked, the assembly System.Private.CoreLib will throw a non-dervied System.IOException. Any other kind of exceptions also these who derives from IOException are forwarded to consumer. This is the only way to get behaviour "When you are was locked" then return false, but when you failed for any other reason, then let consumer handle this [error](https://docs.microsoft.com/en-US/dotnet/api/system.io.file.open?view=netcore-3.1).